## Homework 6 [50-POINTS]

### Statement of Non-cheating and honest collaboration:

I have been informed of the academic honesty policies of Georgia State University. I will abide by those policies. I have written up the solutions below by myself and understand what I have turned in. I collaborated and discussed the problems with the following people: ```[insert here]```

Signed,
```Jesse Annan```


### Directions ** IMPORTANT**

- **You are required to upload all data and code used as a zip file.**
- **Change the notebook's name and add your name before submitting.**
- **You can either code the solution or present your manual calculation. Using inbuilt functions for max, min, sort, sum, etc are OK but using functions for decisiontreeclassifier, kneighborsclassifier, etc are NOT OK**

In [308]:
import numpy as np
import pandas as pd
from numpy.linalg import norm

#### Q-1)[5-Points] The table below lists a dataset that was used to create a nearest neighbor model that predicts whether it will be a good day to go surfing.
|ID |Wave Size (FT)| Web Period (SECS)| Wind Speed (MPH)| Good Surf|
|---|---|---|---|---|
|1|6|15|5|yes|
|2|1|6|9|no|
|3|7|10|4|yes|
|4|7|12|3|yes|
|5|2|2|10|no|
|6|10|2|20|no|

#### Assuming that the model uses Euclidean distance to find the nearest neighbor, what prediction will the model return for each of the following query instances?
|ID |Wave Size (FT)| Web Period (SECS)| Wind Speed (MPH)| Good Surf|
|---|---|---|---|---|
|Q1|8|15|2|?|
|Q2|8|2|18|?|
|Q3|6|12|4|?|

In [309]:
df = pd.read_csv('ques1.csv', sep=r',')
qdf = pd.read_csv('ques1un.csv', sep=r',')
features = ['Wave Size (FT)', 'Web Period (SECS)', 'Wind Speed (MPH)']
df

ID  Wave Size (FT)  Web Period (SECS)  Wind Speed (MPH) Good Surf
0   1               6                 15                 5       yes
1   2               1                  6                 9        no
2   3               7                 10                 4       yes
3   4               7                 12                 3       yes
4   5               2                  2                10        no
5   6              10                  2                20        no

In [310]:
def eucdist(df, row):
    edist = (df - row)**2
    edist = np.sqrt( edist.apply(np.sum, axis=1).to_list() )
    return edist

for k in range(qdf.shape[0]):
    tag = f'eucdist Q{k+1}'
    d = eucdist(df[features], qdf[features].iloc[k])
    df[tag] = d

df

ID  Wave Size (FT)  Web Period (SECS)  Wind Speed (MPH) Good Surf  \
0   1               6                 15                 5       yes   
1   2               1                  6                 9        no   
2   3               7                 10                 4       yes   
3   4               7                 12                 3       yes   
4   5               2                  2                10        no   
5   6              10                  2                20        no   

   eucdist Q1  eucdist Q2  eucdist Q3  
0    3.605551   18.493242    3.162278  
1   13.379088   12.083046    9.273618  
2    5.477226   16.155494    2.236068  
3    3.316625   18.055470    1.414214  
4   16.401219   10.000000   12.328828  
5   22.293497    2.828427   19.287302

In [311]:
colns = ['eucdist Q1','eucdist Q2','eucdist Q3']
prediction = []

for coln in colns:
    mv = np.min(df[coln]) # minimum distance in the column
    idmv = df[coln].to_list().index(mv) # index of min value in the column
    prediction.append(df.loc[idmv,'Good Surf'])

prediction
qdf['Good Surf'] = prediction
qdf 

ID  Wave Size (FT)  Web Period (SECS)  Wind Speed (MPH) Good Surf
0  Q1               8                 15                 2       yes
1  Q2               8                  2                18        no
2  Q3               6                 12                 4       yes

#### Q-2)[35-POINTS] The predictive task in this question is to predict the level of corruption in a country based on a range of macroeconomic and social features. The table below lists some countries described by the following descriptive features: 
- LIFE EXP., the mean life expectancy at birth
- TOP-10 INCOME, the percentage of the annual income of the country that goes to the top 10% of earners
- INFANT MORT., the number of infant deaths per 1,000 births
- MIL. SPEND, the percentage of GDP spent on the military
- SCHOOL YEARS, the mean number years spent in school by adult females

#### The target feature is the Corruption Perception Index (CPI). The CPI measures the perceived levels of corruption in the public sector of countries and ranges from 0(highly corrupt) to 100 (very clean).<sup>1</sup>

|COUNTRY ID |LIFE EXP.| TOP-10 INCOME| INFANT MORT.| Mil. SPEND| SCHOOL YEARS|CPI|
|---|---|---|---|---|---|---|
|Afghanistan|59.61|23.21|74.30|4.44|0.40|1.5171|
|Haiti|45.00|47.67|73.10|0.09|3.40|1.7999|
|Nigeria|51.30|38.23|82.60|1.07|4.10|2.4493|
|Egypt| 70.48| 26.58|19.60|1.86|5.30|2.8622|
|Argentina|75.77|32.30|13.30| 0.76|10.10|2.9961|
|China|74.87|29.98|13.70|1.95|6.40|3.6356|
|Brazil|73.12|42.93|14.50|1.43|7.20|3.7741|
|Israel|81.30|28.80|3.60|6.77|12.50|5.8069|
|USA|78.51|29.85|6.30|4.72|13.70|7.1357|
|Ireland| 80.15|27.23|3.50|0.60|11.50|7.5360|
|UK|80.09|28.49|4.40|2.59|13.00|7.7751|
|Germany|80.24|22.07|3.50|1.31|12.00|8.0461|
|Canada|80.99| 24.79|4.90|1.42|14.20|8.6725|
|Australia| 82.09|25.40|4.20|1.86|11.50|8.8442|
|Sweden|81.43|22.18|2.40|1.27|12.80|9.2985|
|New Zealand| 80.67|27.81|4.90|1.13|12.30|9.4627|

#### We will use Russia as our query country for this question. The table below lists the descriptive features for Russia.

|COUNTRY ID |LIFE EXP.| TOP-10 INCOME| INFANT MORT.| Mil. SPEND| SCHOOL YEARS|CPI|
|---|---|---|---|---|---|---|
|Russia|67.62|31.68|10.00|3.87|12.90|?|
###

    a)[5-Points] What value would a 3-nearest neighbor prediction model using Euclidean distance return for the CPI of Russia?

    b)[5-Points] What value would a weighted k-NN prediction model return for the CPI of Russia? Use k “ 16 (i.e., the full dataset) and a weighting scheme of the reciprocal of the squared Euclidean distance between the neighbor and the query.

    c)[5-Points] What value would a k-NN model with k = 3 and using Manhattan distance return for the CPI of Russia?

    d)[5-Points] What value would a 3-NN model using cosine similarity return for the CPI of Russia?

    e)[6-Points] The descriptive features in this dataset are of different types. For example, some are percentages, others are measured in years, and others are measured in counts per 1,000. We should always consider normalizing our data, but it is particularly important to do this when the descriptive features are measured in different units. What value would a 3-nearest neighbor prediction model using Euclidean distance return for the CPI of Russia when the descriptive features have been normalized using range normalization?
    
    f)[7-Points] What value would a weighted k-NN prediction model—with k = 16 (i.e., the full dataset) and using a weighting scheme of the reciprocal of the squared Euclidean distance between the neighbor and the query—return for the CPI of Russia when it is applied to the range-normalized data?
    
    g)[2-Points] The actual 2011 CPI for Russia was 2.4488. Which of the predictions made was the most accurate? Why do you think this was?

NOTE: 1. The data listed in this table is real and is for 2010/11 (or the most recent year prior to 2010/11 when the data was available). The data for the descriptive features in this table was amalgamated from a number of surveys retrieved from Gapminder (www.gapminder.org). The Corruption Perception Index is generated annually by Transparency International (www.transparency.org).

In [312]:
countdf = pd.read_csv('ques2.csv', sep=r',')
qcountdf = pd.read_csv('ques2un.csv', sep=r',')
features = countdf.columns.to_list()[1:-1]
countdf

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0   Afghanistan      59.61          23.21          74.3        4.44   
1         Haiti      45.00          47.67          73.1        0.09   
2       Nigeria      51.30          38.23          82.6        1.07   
3         Egypt      70.48          26.58          19.6        1.86   
4     Argentina      75.77          32.30          13.3        0.76   
5         China      74.87          29.98          13.7        1.95   
6        Brazil      73.12          42.93          14.5        1.43   
7        Israel      81.30          28.80           3.6        6.77   
8           USA      78.51          29.85           6.3        4.72   
9       Ireland      80.15          27.23           3.5        0.60   
10           UK      80.09          28.49           4.4        2.59   
11      Germany      80.24          22.07           3.5        1.31   
12       Canada      80.99          24.79           4.9        1.42   
13    Australia      82.09          25.40           4.2        1.86   
14       Sweden      81.43          22.18           2.4        1.27   
15  New Zealand      80.67          27.81           4.9        1.13   

    SCHOOL YEARS     CPI  
0            0.4  1.5171  
1            3.4  1.7999  
2            4.1  2.4493  
3            5.3  2.8622  
4           10.1  2.9961  
5            6.4  3.6356  
6            7.2  3.7741  
7           12.5  5.8069  
8           13.7  7.1357  
9           11.5  7.5360  
10          13.0  7.7751  
11          12.0  8.0461  
12          14.2  8.6725  
13          11.5  8.8442  
14          12.8  9.2985  
15          12.3  9.4627

#### [a]

In [313]:
countdf_a = countdf.copy()
qcountdf_a = qcountdf.copy()

# Euclidian distance between each country and Russia
dist = eucdist(countdf_a[features], qcountdf_a[features].iloc[0])
countdf_a['eucdist Rus'] = dist
countdf_a.head()

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0  Afghanistan      59.61          23.21          74.3        4.44   
1        Haiti      45.00          47.67          73.1        0.09   
2      Nigeria      51.30          38.23          82.6        1.07   
3        Egypt      70.48          26.58          19.6        1.86   
4    Argentina      75.77          32.30          13.3        0.76   

   SCHOOL YEARS     CPI  eucdist Rus  
0           0.4  1.5171    66.535448  
1           3.4  1.7999    69.667014  
2           4.1  2.4493    75.268087  
3           5.3  2.8622    13.716767  
4          10.1  2.9961     9.757510

In [314]:
k = 3
near_neighbors = []
sorted_eucdist = countdf_a['eucdist Rus'].sort_values().to_list()

for k in range(k):
    for val in countdf_a['eucdist Rus']:
        if sorted_eucdist[k] == val:
            idx = countdf_a['eucdist Rus'].to_list().index(sorted_eucdist[k])
            near_neighbors.append(countdf_a.loc[idx,'CPI'])
            # print(countdf_a.iloc[[idx]]) # neareast neighbours

prediction = np.mean(near_neighbors)
qcountdf_a['CPI'] = prediction
qcountdf_a

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0     Russia      67.62          31.68          10.0        3.87   

   SCHOOL YEARS       CPI  
0          12.9  4.589133

#### [b]

In [315]:
countdf_b = countdf.copy()
qcountdf_b = qcountdf.copy()

# Resciprocal of squared euclidian distance between each country and Russia
def weucdist(df, row):
    wedist = 1. / (df - row)**2
    wedist = wedist.apply(np.sum, axis=1).to_list()
    return wedist

dist = weucdist(countdf_b[features], qcountdf_b[features].iloc[0])
countdf_b['weucdist Rus'] = dist
countdf_b.head()


COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0  Afghanistan      59.61          23.21          74.3        4.44   
1        Haiti      45.00          47.67          73.1        0.09   
2      Nigeria      51.30          38.23          82.6        1.07   
3        Egypt      70.48          26.58          19.6        1.86   
4    Argentina      75.77          32.30          13.3        0.76   

   SCHOOL YEARS     CPI  weucdist Rus  
0           0.4  1.5171      3.114037  
1           3.4  1.7999      0.087184  
2           4.1  2.4493      0.167717  
3           5.3  2.8622      0.436384  
4          10.1  2.9961      2.939281

In [316]:
prediction = countdf_b['CPI'].mean()
qcountdf_b['CPI'] = prediction
qcountdf_b

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0     Russia      67.62          31.68          10.0        3.87   

   SCHOOL YEARS      CPI  
0          12.9  5.72575

#### [c]

In [317]:
countdf_c = countdf.copy()
qcountdf_c = qcountdf.copy()

def manhdist(df, row):
    mdist = np.abs(df - row)
    mdist = mdist.apply(np.sum, axis=1).to_list()
    return mdist

# Manhattan distance between each country and Russia
dist = manhdist(countdf_c[features], qcountdf_c[features].iloc[0])
countdf_c['manhdist Rus'] = dist
countdf_c.head()

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0  Afghanistan      59.61          23.21          74.3        4.44   
1        Haiti      45.00          47.67          73.1        0.09   
2      Nigeria      51.30          38.23          82.6        1.07   
3        Egypt      70.48          26.58          19.6        1.86   
4    Argentina      75.77          32.30          13.3        0.76   

   SCHOOL YEARS     CPI  manhdist Rus  
0           0.4  1.5171         93.85  
1           3.4  1.7999        114.99  
2           4.1  2.4493        107.07  
3           5.3  2.8622         27.17  
4          10.1  2.9961         17.98

In [318]:
k = 3
near_neighbors = []
sorted_eucdist = countdf_c['manhdist Rus'].sort_values().to_list()

for k in range(k):
    for val in countdf_c['manhdist Rus']:
        if sorted_eucdist[k] == val:
            idx = countdf_c['manhdist Rus'].to_list().index(sorted_eucdist[k])
            near_neighbors.append(countdf_c.loc[idx,'CPI'])
            # print(countdf_c.iloc[[idx]]) # neareast neighbours

prediction = np.mean(near_neighbors)
qcountdf_c['CPI'] = prediction
qcountdf_c

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0     Russia      67.62          31.68          10.0        3.87   

   SCHOOL YEARS       CPI  
0          12.9  4.589133

#### [d]

In [319]:
countdf_d = countdf.copy()
qcountdf_d = qcountdf.copy()

def cosinedist(df, rus):
    cosinedist = []
    for _, row in countdf_d.iterrows():
        row = row[features]
        cdist = np.dot(row, rus)/(norm(row)*norm(rus))
        cosinedist.append(cdist)
    return cosinedist

# Cosine distance between each country and Russia
dist = cosinedist(countdf_d[features], qcountdf_d[features].iloc[0].to_list())
countdf_d['cosinedist Rus'] = dist
countdf_d.head()

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0  Afghanistan      59.61          23.21          74.3        4.44   
1        Haiti      45.00          47.67          73.1        0.09   
2      Nigeria      51.30          38.23          82.6        1.07   
3        Egypt      70.48          26.58          19.6        1.86   
4    Argentina      75.77          32.30          13.3        0.76   

   SCHOOL YEARS     CPI  cosinedist Rus  
0           0.4  1.5171        0.736406  
1           3.4  1.7999        0.708616  
2           4.1  2.4493        0.695137  
3           5.3  2.8622        0.984437  
4          10.1  2.9961        0.996986

In [320]:
k = 3
near_neighbors = []
sorted_eucdist = countdf_d['cosinedist Rus'].sort_values().to_list()

for k in range(k):
    for val in countdf_d['cosinedist Rus']:
        if sorted_eucdist[k] == val:
            idx = countdf_d['cosinedist Rus'].to_list().index(sorted_eucdist[k])
            near_neighbors.append(countdf_d.loc[idx,'CPI'])
            # print(countdf_d.iloc[[idx]]) # neareast neighbours

prediction = np.mean(near_neighbors)
qcountdf_d['CPI'] = prediction
qcountdf_d

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0     Russia      67.62          31.68          10.0        3.87   

   SCHOOL YEARS     CPI  
0          12.9  1.9221

#### [e]

In [321]:
countdf_e = countdf.copy()
qcountdf_e = qcountdf.copy()
df = pd.concat([countdf_e, qcountdf_e], ignore_index=1)

minmax_norm = lambda col: (col - col.min()) / (col.max() - col.min())
df[features] = df[features].apply(minmax_norm)
countdf_e, qcountdf_e = df.drop([16]), df.iloc[[16]]

dist = eucdist(countdf_e[features], qcountdf_e[features].iloc[0])
countdf_e['eucdist Rus'] = dist
countdf_e.head()

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0  Afghanistan   0.393907       0.044531      0.896509    0.651198   
1        Haiti   0.000000       1.000000      0.881546    0.000000   
2      Nigeria   0.169857       0.631250      1.000000    0.146707   
3        Egypt   0.686978       0.176172      0.214464    0.264970   
4    Argentina   0.829604       0.399609      0.135910    0.100299   

   SCHOOL YEARS     CPI  eucdist Rus  
0      0.000000  1.5171     1.275403  
1      0.217391  1.7999     1.474860  
2      0.268116  2.4493     1.288744  
3      0.355072  2.8622     0.673647  
4      0.702899  2.9961     0.555415

In [322]:
k = 3
near_neighbors = []
sorted_eucdist = countdf_e['eucdist Rus'].sort_values().to_list()

for k in range(k):
    for val in countdf_e['eucdist Rus']:
        if sorted_eucdist[k] == val:
            idx = countdf_e['eucdist Rus'].to_list().index(sorted_eucdist[k])
            near_neighbors.append(countdf_e.loc[idx,'CPI'])
            # print(countdf_e.iloc[[idx]]) # neareast neighbours

prediction = np.mean(near_neighbors)
qcountdf_e.loc[16,'CPI'] = prediction
qcountdf_e

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
16     Russia   0.609868       0.375391      0.094763    0.565868   

    SCHOOL YEARS       CPI  
16      0.905797  5.968967

#### [f]

In [323]:
countdf_f = countdf.copy()
qcountdf_f = qcountdf.copy()
tempdf = pd.concat([countdf_f, qcountdf_f], ignore_index=1)

tempdf[features] = tempdf[features].apply(minmax_norm)
countdf_f, qcountdf_f = tempdf.drop([16]), tempdf.iloc[[16]]

dist = weucdist(countdf_f[features], qcountdf_f[features].iloc[0])
countdf_f['weucdist Rus'] = dist
countdf_f.head()

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
0  Afghanistan   0.393907       0.044531      0.896509    0.651198   
1        Haiti   0.000000       1.000000      0.881546    0.000000   
2      Nigeria   0.169857       0.631250      1.000000    0.146707   
3        Egypt   0.686978       0.176172      0.214464    0.264970   
4    Argentina   0.829604       0.399609      0.135910    0.100299   

   SCHOOL YEARS     CPI  weucdist Rus  
0      0.000000  1.5171    170.692744  
1      0.217391  1.7999     12.100373  
2      0.268116  2.4493     29.811755  
3      0.355072  2.8622    277.513339  
4      0.702899  2.9961   2345.143229

In [324]:
prediction = countdf_f['CPI'].mean()
qcountdf_f.loc[16,'CPI'] = prediction
qcountdf_f

COUNTRY ID  LIFE EXP.  TOP-10 INCOME  INFANT MORT.  Mil. SPEND  \
16     Russia   0.609868       0.375391      0.094763    0.565868   

    SCHOOL YEARS      CPI  
16      0.905797  5.72575

#### [g]
Cosine similarity performed better because it does not take into consideration the distance between the vectors but rather the angle between vectors. 

#### Q-3)[10] You have been given the job of building a recommender system for a large online shop that has a stock of over 100,000 items. In this domain the behavior of customers is captured in terms of what items they have bought or not bought. For example, the following table lists the behavior of two customers in this domain for a subset of the items that at least one of the customers has bought.

| ID| ITEM 107| ITEM 498| ITEM 7256| ITEM 28063| ITEM 75328|
|---|---|---|---|---|---|
| 1| true| true| true| false| false|
| 2| true| false| flase| true| true|

    a)[5-Points] The company has decided to use a similarity-based model to implement the recommender system. Which of the following three similarity indexes do you think the system should be based on?
    
    Russel-Rao(X,Y) = CP(X,Y)/P
    Sokal-Michener(X,Y) = (CP(X,Y) + CA(X,Y))/P
    Jaccard(X,Y) = CP(X,Y)/(CP(X,Y) + PA(X,Y) + AP(X,Y))

    b)[5-Points] What items will the system recommend to the following customer? Assume that the recommender system uses the similarity index you chose in the first part of this question and is trained on the sample dataset listed above. Also assume that the system generates recommendations for query customers by finding the customer most similar to them in the dataset and then recommending the items that this similar customer has bought but that the query customer has not bought.
| ID| ITEM 107| ITEM 498| ITEM 7256| ITEM 28063| ITEM 75328|
|---|---|---|---|---|---|
| Query| true| flase| true| false| false|

In [325]:
# Similarity between D1 and Query (D1, Q)
CP = 2
CA = 2
PA = 1
AP = 0
P = CP + CA + PA + AP

russel_r = CP / P
sokal_m = (CP + CA) / P
jaccard = CP / (P - CA)

print(f'Russel-Rao(D1,Query) = {russel_r} \nSokal-Michener(D1,Query) = {sokal_m} \nJaccard(D1,Query) = {jaccard}')


Russel-Rao(D1,Query) = 0.4 
Sokal-Michener(D1,Query) = 0.8 
Jaccard(D1,Query) = 0.6666666666666666


In [326]:
# Similarity between D2 and Query
CP = 1
CA = 1
PA = 2
AP = 1
P = CP + CA + PA + AP

russel_r = CP / P
sokal_m = (CP + CA) / P
jaccard = CP / (P - CA)

print(f'Russel-Rao(D1,Query) = {russel_r} \nSokal-Michener(D1,Query) = {sokal_m} \nJaccard(D1,Query) = {jaccard}')


Russel-Rao(D1,Query) = 0.2 
Sokal-Michener(D1,Query) = 0.4 
Jaccard(D1,Query) = 0.25


#### [a]
Russel-Rao(X,Y) = CP(X,Y)/P

#### [b]
Based the Russel-Rao metric, `D1` is judged to be more similar to the `Query`. Therefore the recommender system will recomment `ITEM 498` to the customer.